### ISSUE TITLE
Wayland protocol requirements

### ISSUE THREAD

**Issue by christianrauch** [glfw/glfw#1637]
The `wayland-protocols` version 1.15 is requested, presumably to access `xdg_decoration_unstable_v1`. Some unstable Wayland protocols should be optional, such as the xdg decorations, to allow the use of older `wayland-protocols` versions.

In general, I think that a higher package version should always be justified by requirements and that new features should be optional, if possible.

**Comment by linkmauve** [comment]
This version of `wayland-protocols` is required at build-time, but this is only in order to generate the C files that allow GLFW to speak these protocols. At runtime [we bind every protocol global we know of](https://github.com/glfw/glfw/blob/master/src/wl_init.c#L783), but aside from [xdg-shell being mandatory](https://github.com/glfw/glfw/blob/master/src/wl_init.c#L1160) to create a window, and the core protocol as well, none of the extensions are required for GLFW to run.

**Comment by christianrauch** [comment]
I understand that these protocol definitions are build-time dependencies and they are checked at runtime anyway. But it still prevents building GLFW on older systems. The `CMakeLists.txt` could simply detect if a definition is available and deactivate their usage, e.g. the `zxdg_decoration_manager_v1`, via ifdefs.

Since the `xdg_shell` protocol is stable for a long time now and supported widely, I do not see an issue with removing the deprecated `wl_shell`.

**Comment by linkmauve** [comment]
The issue with that approach is that it would introduce a lot of complexity in the code by littering it with `#ifdef` everywhere. Also the `xdg-decoration` protocol was released one and a half years ago, just six months after `xdg-shell`.

You should be able to update `wayland-protocols` very easily no matter your distribution. It’s a simple repository of XML files, no building is even required you can just `git clone` it and be done with it, request your distribution to provide a newer version in their backports, or vendor it with your application.

**Comment by christianrauch** [comment]
Well, would you accept a PR that implements that ifdef compiler switch?

**Comment by elmindreda** [comment]
I agree with @linkmauve that we shouldn't introduce conditional compilation for this. Wayland is changing and growing fast enough as it is without adding more complexity into that backend. For backward compatibility on the Linux desktop we have the X11 backend.
